File: java/lib/src/main/java/com/svix/api/Authentication.java
// this file is @generated
package com.svix.api;

import com.svix.SvixHttpClient;
import com.svix.exceptions.ApiException;
import com.svix.models.ApiTokenOut;
import com.svix.models.AppPortalAccessIn;
import com.svix.models.AppPortalAccessOut;
import com.svix.models.ApplicationTokenExpireIn;
import com.svix.models.RotatePollerTokenIn;
import com.svix.models.StreamPortalAccessIn;

import okhttp3.Headers;
import okhttp3.HttpUrl;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class Authentication {
    private final SvixHttpClient client;

    public Authentication(SvixHttpClient client) {
        this.client = client;
    }

    /**
     * Use this function to get magic links (and authentication codes) for connecting your users to
     * the Consumer Application Portal.
     */
    public AppPortalAccessOut appPortalAccess(
            final String appId, final AppPortalAccessIn appPortalAccessIn)
            throws IOException, ApiException {
        return this.appPortalAccess(
                appId, appPortalAccessIn, new AuthenticationAppPortalAccessOptions());
    }

    /**
     * Use this function to get magic links (and authentication codes) for connecting your users to
     * the Consumer Application Portal.
     */
    public AppPortalAccessOut appPortalAccess(
            final String appId,
            final AppPortalAccessIn appPortalAccessIn,
            final AuthenticationAppPortalAccessOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/auth/app-portal-access/%s", appId));
        Map<String, String> headers = new HashMap<>();
        if (options.idempotencyKey != null) {
            headers.put("idempotency-key", options.idempotencyKey);
        }
        return this.client.executeRequest(
                "POST",
                url.build(),
                Headers.of(headers),
                appPortalAccessIn,
                AppPortalAccessOut.class);
    }

    /** Expire all of the tokens associated with a specific application. */
    public void expireAll(
            final String appId, final ApplicationTokenExpireIn applicationTokenExpireIn)
            throws IOException, ApiException {
        this.expireAll(appId, applicationTokenExpireIn, new AuthenticationExpireAllOptions());
    }

    /** Expire all of the tokens associated with a specific application. */
    public void expireAll(
            final String appId,
            final ApplicationTokenExpireIn applicationTokenExpireIn,
            final AuthenticationExpireAllOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/auth/app/%s/expire-all", appId));
        Map<String, String> headers = new HashMap<>();
        if (options.idempotencyKey != null) {
            headers.put("idempotency-key", options.idempotencyKey);
        }
        this.client.executeRequest(
                "POST", url.build(), Headers.of(headers), applicationTokenExpireIn, null);
    }

    /**
     * @deprecated Please use appPortalAccess instead.
     */
    @Deprecated
    public com.svix.models.DashboardAccessOut dashboardAccess(final String appId)
            throws IOException, ApiException {
        return this.dashboardAccess(appId, new AuthenticationDashboardAccessOptions());
    }

    /**
     * @deprecated Please use appPortalAccess instead.
     */
    @Deprecated
    public com.svix.models.DashboardAccessOut dashboardAccess(
            final String appId, final AuthenticationDashboardAccessOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/auth/dashboard-access/%s", appId));
        Map<String, String> headers = new HashMap<>();
        if (options.idempotencyKey != null) {
            headers.put("idempotency-key", options.idempotencyKey);
        }
        return this.client.executeRequest(
                "POST",
                url.build(),
                Headers.of(headers),
                null,
                com.svix.models.DashboardAccessOut.class);
    }

    /**
     * Logout an app token.
     *
     * <p>Trying to log out other tokens will fail.
     */
    public void logout() throws IOException, ApiException {

        this.logout(new AuthenticationLogoutOptions());
    }

    /**
     * Logout an app token.
     *
     * <p>Trying to log out other tokens will fail.
     */
    public void logout(final AuthenticationLogoutOptions options) throws IOException, ApiException {
        HttpUrl.Builder url = this.client.newUrlBuilder().encodedPath("/api/v1/auth/logout");
        Map<String, String> headers = new HashMap<>();
        if (options.idempotencyKey != null) {
            headers.put("idempotency-key", options.idempotencyKey);
        }
        this.client.executeRequest("POST", url.build(), Headers.of(headers), null, null);
    }

    /**
     * Use this function to get magic links (and authentication codes) for connecting your users to
     * the Stream Consumer Portal.
     */
    public AppPortalAccessOut streamPortalAccess(
            final String streamId, final StreamPortalAccessIn streamPortalAccessIn)
            throws IOException, ApiException {
        return this.streamPortalAccess(
                streamId, streamPortalAccessIn, new AuthenticationStreamPortalAccessOptions());
    }

    /**
     * Use this function to get magic links (and authentication codes) for connecting your users to
     * the Stream Consumer Portal.
     */
    public AppPortalAccessOut streamPortalAccess(
            final String streamId,
            final StreamPortalAccessIn streamPortalAccessIn,
            final AuthenticationStreamPortalAccessOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(
                                String.format("/api/v1/auth/stream-portal-access/%s", streamId));
        Map<String, String> headers = new HashMap<>();
        if (options.idempotencyKey != null) {
            headers.put("idempotency-key", options.idempotencyKey);
        }
        return this.client.executeRequest(
                "POST",
                url.build(),
                Headers.of(headers),
                streamPortalAccessIn,
                AppPortalAccessOut.class);
    }

    /** Get the current auth token for the stream poller. */
    public ApiTokenOut getStreamPollerToken(final String streamId, final String sinkId)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(
                                String.format(
                                        "/api/v1/auth/stream/%s/sink/%s/poller/token",
                                        streamId, sinkId));
        return this.client.executeRequest("GET", url.build(), null, null, ApiTokenOut.class);
    }

    /** Create a new auth token for the stream poller API. */
    public ApiTokenOut rotateStreamPollerToken(
            final String streamId,
            final String sinkId,
            final RotatePollerTokenIn rotatePollerTokenIn)
            throws IOException, ApiException {
        return this.rotateStreamPollerToken(
                streamId,
                sinkId,
                rotatePollerTokenIn,
                new AuthenticationRotateStreamPollerTokenOptions());
    }

    /** Create a new auth token for the stream poller API. */
    public ApiTokenOut rotateStreamPollerToken(
            final String streamId,
            final String sinkId,
            final RotatePollerTokenIn rotatePollerTokenIn,
            final AuthenticationRotateStreamPollerTokenOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(
                                String.format(
                                        "/api/v1/auth/stream/%s/sink/%s/poller/token/rotate",
                                        streamId, sinkId));
        Map<String, String> headers = new HashMap<>();
        if (options.idempotencyKey != null) {
            headers.put("idempotency-key", options.idempotencyKey);
        }
        return this.client.executeRequest(
                "POST", url.build(), Headers.of(headers), rotatePollerTokenIn, ApiTokenOut.class);
    }
}

File: kotlin/lib/src/main/kotlin/Authentication.kt
// this file is @generated
package com.svix.kotlin

import com.svix.kotlin.models.ApiTokenOut
import com.svix.kotlin.models.AppPortalAccessIn
import com.svix.kotlin.models.AppPortalAccessOut
import com.svix.kotlin.models.ApplicationTokenExpireIn
import com.svix.kotlin.models.RotatePollerTokenIn
import com.svix.kotlin.models.StreamPortalAccessIn
import okhttp3.Headers

data class AuthenticationAppPortalAccessOptions(val idempotencyKey: String? = null)

data class AuthenticationExpireAllOptions(val idempotencyKey: String? = null)

data class AuthenticationLogoutOptions(val idempotencyKey: String? = null)

data class AuthenticationStreamPortalAccessOptions(val idempotencyKey: String? = null)

data class AuthenticationRotateStreamPollerTokenOptions(val idempotencyKey: String? = null)

class Authentication(private val client: SvixHttpClient) {
    /**
     * Use this function to get magic links (and authentication codes) for connecting your users to
     * the Consumer Application Portal.
     */
    suspend fun appPortalAccess(
        appId: String,
        appPortalAccessIn: AppPortalAccessIn,
        options: AuthenticationAppPortalAccessOptions = AuthenticationAppPortalAccessOptions(),
    ): AppPortalAccessOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/auth/app-portal-access/$appId")
        val headers = Headers.Builder()
        options.idempotencyKey?.let { headers.add("idempotency-key", it) }

        return client.executeRequest<AppPortalAccessIn, AppPortalAccessOut>(
            "POST",
            url.build(),
            headers = headers.build(),
            reqBody = appPortalAccessIn,
        )
    }

    /** Expire all of the tokens associated with a specific application. */
    suspend fun expireAll(
        appId: String,
        applicationTokenExpireIn: ApplicationTokenExpireIn,
        options: AuthenticationExpireAllOptions = AuthenticationExpireAllOptions(),
    ) {
        val url = client.newUrlBuilder().encodedPath("/api/v1/auth/app/$appId/expire-all")
        val headers = Headers.Builder()
        options.idempotencyKey?.let { headers.add("idempotency-key", it) }

        client.executeRequest<ApplicationTokenExpireIn, Boolean>(
            "POST",
            url.build(),
            headers = headers.build(),
            reqBody = applicationTokenExpireIn,
        )
    }

    /** @deprecated Please use `appPortalAccess` instead. */
    @Deprecated("Please use `appPortalAccess` instead.")
    suspend fun dashboardAccess(
        appId: String,
        options: AuthenticationDashboardAccessOptions = AuthenticationDashboardAccessOptions(),
    ): com.svix.kotlin.models.DashboardAccessOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/auth/dashboard-access/$appId")
        val headers = Headers.Builder()
        options.idempotencyKey?.let { headers.add("idempotency-key", it) }
        return client.executeRequest<Any, com.svix.kotlin.models.DashboardAccessOut>(
            "POST",
            url.build(),
            headers = headers.build(),
        )
    }

    /**
     * Logout an app token.
     *
     * Trying to log out other tokens will fail.
     */
    suspend fun logout(options: AuthenticationLogoutOptions = AuthenticationLogoutOptions()) {
        val url = client.newUrlBuilder().encodedPath("/api/v1/auth/logout")
        val headers = Headers.Builder()
        options.idempotencyKey?.let { headers.add("idempotency-key", it) }
        client.executeRequest<Any, Boolean>("POST", url.build(), headers = headers.build())
    }

    /**
     * Use this function to get magic links (and authentication codes) for connecting your users to
     * the Stream Consumer Portal.
     */
    suspend fun streamPortalAccess(
        streamId: String,
        streamPortalAccessIn: StreamPortalAccessIn,
        options: AuthenticationStreamPortalAccessOptions = AuthenticationStreamPortalAccessOptions(),
    ): AppPortalAccessOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/auth/stream-portal-access/$streamId")
        val headers = Headers.Builder()
        options.idempotencyKey?.let { headers.add("idempotency-key", it) }

        return client.executeRequest<StreamPortalAccessIn, AppPortalAccessOut>(
            "POST",
            url.build(),
            headers = headers.build(),
            reqBody = streamPortalAccessIn,
        )
    }

    /** Get the current auth token for the stream poller. */
    suspend fun getStreamPollerToken(streamId: String, sinkId: String): ApiTokenOut {
        val url =
            client
                .newUrlBuilder()
                .encodedPath("/api/v1/auth/stream/$streamId/sink/$sinkId/poller/token")
        return client.executeRequest<Any, ApiTokenOut>("GET", url.build())
    }

    /** Create a new auth token for the stream poller API. */
    suspend fun rotateStreamPollerToken(
        streamId: String,
        sinkId: String,
        rotatePollerTokenIn: RotatePollerTokenIn,
        options: AuthenticationRotateStreamPollerTokenOptions =
            AuthenticationRotateStreamPollerTokenOptions(),
    ): ApiTokenOut {
        val url =
            client
                .newUrlBuilder()
                .encodedPath("/api/v1/auth/stream/$streamId/sink/$sinkId/poller/token/rotate")
        val headers = Headers.Builder()
        options.idempotencyKey?.let { headers.add("idempotency-key", it) }

        return client.executeRequest<RotatePollerTokenIn, ApiTokenOut>(
            "POST",
            url.build(),
            headers = headers.build(),
            reqBody = rotatePollerTokenIn,
        )
    }
}
